toolchain/lower/function_context.cpp

```diff
@@ -37,19 +37,33 @@ auto FunctionContext::TryToReuseBlock(SemIR::InstBlockId block_id,
   return true;
 }
 
+// Structure for the core handler dispatch.
+using DispatchFunctionT = auto(FunctionContext& context, SemIR::InstId inst_id,
+                               SemIR::Inst inst) -> void;
+using DispatchTableT =
+    std::array<DispatchFunctionT*, SemIR::InstKind::EnumCount>;
+
+// Transforms a parse node ID to a typed parse node, which is then passed to the
+// handler.
+#define CARBON_SEM_IR_INST_KIND(Name)                                  \
+  auto Dispatch##Name(FunctionContext& context, SemIR::InstId inst_id, \
+                      SemIR::Inst inst) -> void {                      \
+    return Handle##Name(context, inst_id, inst.As<SemIR::Name>());     \
+  }
+#include "toolchain/sem_ir/inst_kind.def"
+
+// The main dispatch table. This is used instead of a switch in order to
+// optimize for common functionality.
+static constexpr DispatchTableT DispatchTable = {
+#define CARBON_SEM_IR_INST_KIND(Name) &Dispatch##Name,
+#include "toolchain/sem_ir/inst_kind.def"
+};
```

Comment:
This enum count / array sizing technique seems to be adding a bit of complexity to the enum definitions that we otherwise haven't needed so far. I'm not opposed, but I wonder if you considered an inferred array bound instead:

```suggestion
// Structure for the core handler dispatch.
using DispatchFunctionT = auto(FunctionContext& context, SemIR::InstId inst_id,
                               SemIR::Inst inst) -> void;

// Transforms a parse node ID to a typed parse node, which is then passed to the
// handler.
#define CARBON_SEM_IR_INST_KIND(Name)                                  \
  auto Dispatch##Name(FunctionContext& context, SemIR::InstId inst_id, \
                      SemIR::Inst inst) -> void {                      \
    return Handle##Name(context, inst_id, inst.As<SemIR::Name>());     \
  }
#include "toolchain/sem_ir/inst_kind.def"

// The main dispatch table. This is used instead of a switch in order to
// optimize for common functionality.
static constexpr DispatchFunctionT* DispatchTable[] = {
#define CARBON_SEM_IR_INST_KIND(Name) &Dispatch##Name,
#include "toolchain/sem_ir/inst_kind.def"
};
```
We could use `std::array` CTAD and also avoid having to write out a function type:
```suggestion
// Transforms a parse node ID to a typed parse node, which is then passed to the
// handler.
#define CARBON_SEM_IR_INST_KIND(Name)                                  \
  auto Dispatch##Name(FunctionContext& context, SemIR::InstId inst_id, \
                      SemIR::Inst inst) -> void {                      \
    return Handle##Name(context, inst_id, inst.As<SemIR::Name>());     \
  }
#include "toolchain/sem_ir/inst_kind.def"

// The main dispatch table. This is used instead of a switch in order to
// optimize for common functionality.
static constexpr std::array DispatchTable = {
#define CARBON_SEM_IR_INST_KIND(Name) &Dispatch##Name,
#include "toolchain/sem_ir/inst_kind.def"
};
```
We could also shorten this a little by generating lambdas inline:
```suggestion
// The main dispatch table. This is used instead of a switch in order to
// optimize for common functionality.
static constexpr std::array DispatchTable = {
#define CARBON_SEM_IR_INST_KIND(Name)                                \
   +[](FunctionContext& context, SemIR::InstId inst_id,              \
       SemIR::Inst inst) -> void {                                   \
      return Handle##Name(context, inst_id, inst.As<SemIR::Name>()); \
   },
#include "toolchain/sem_ir/inst_kind.def"
};
```

I'm not going to push for a change here if you prefer it as-is.

Reply:
I'm keeping the separate `Dispatch` for now, I think it's a little easier to read that way. But switched to `std::array`